### ISSUE TITLE
CPU Spiking, Needs Caching

### ISSUE THREAD

**Issue by Fluxty** [greatman/craftconomy3#105]
Plugin needs caching methods, causes massive lag spikes on populous servers. Thanks <3

http://i.imgur.com/62EGwvZ.png

http://i.imgur.com/IT5EZYx.png

**Comment by Phoenix616** [comment]
Implementing caching with an economy plugin that is supposed to not have any exploits is really difficult to do, especially in a time where you want economies to work on multiple servers connected to a Bungee network without balancing desyncing.

My suggestion to you would be to put a faster database behind it.

**Comment by Fluxty** [comment]
@Phoenix616 
"net.brcdev.shopgui" is using Vault to check the balance of a player in the main thread. Craftconomy3 ("com.greatmancode.craftconomy3") performs a database query in the main thread in response to the Vault API request.

I'd say Craftconomy is at fault for doing database queries in the main thread, it should cache player balances in memory.

**Comment by Phoenix616** [comment]
@Fluxty Yeah, optional caching would be one way to work around a slow database. If the only queries that you are worried about happen very often and are only balance checks you could even let your database cache the result by itself.

Another one would be for plugins to not access an economy plugin on the main thread. (That could result in new desync-issues 'though as two threads might access the same account at the same time)
